File: backend_samples/library/dotnet/AzureAIVisionFaceDeviceAttestation/src/Services/ApiTelemetry.cs
using Azure.AI.Vision.Face.DeviceAttestation.Logging;

namespace Azure.AI.Vision.Face.DeviceAttestation.Services;

/// <summary>
/// Per-route failure telemetry. Emits a structured "Api.Fail" event so failure
/// reasons can be grouped/queried by route + reason code.
/// </summary>
internal static class ApiTelemetry
{
    public static void TrackApiFail(
        IAttestationLogger logger,
        string route,
        string reason,
        int status,
        IReadOnlyDictionary<string, object?>? properties = null)
    {
        var props = new Dictionary<string, object?>
        {
            ["route"] = route,
            ["reason"] = reason,
            ["status"] = status,
        };
        if (properties is not null)
        {
            foreach (var kv in properties)
            {
                props[kv.Key] = kv.Value;
            }
        }
        logger.TrackEvent("Api.Fail", props);
    }
}

File: backend_samples/library/java/AzureAIVisionFaceDeviceAttestation/src/main/java/com/azure/ai/vision/face/deviceattestation/services/ApiTelemetry.java
package com.azure.ai.vision.face.deviceattestation.services;

import java.util.HashMap;
import java.util.Map;

import com.azure.ai.vision.face.deviceattestation.logging.AttestationLogger;

/**
 * Per-route failure telemetry. Emits a structured "Api.Fail" event so failure
 * reasons can be grouped/queried by route + reason code.
 */
public final class ApiTelemetry {

    public static void trackApiFail(
            AttestationLogger logger,
            String route,
            String reason,
            int status,
            Map<String, Object> properties) {
        Map<String, Object> props = new HashMap<>();
        props.put("route", route);
        props.put("reason", reason);
        props.put("status", status);
        if (properties != null) {
            props.putAll(properties);
        }
        logger.trackEvent("Api.Fail", props, null);
    }

    private ApiTelemetry() {
    }
}
